[AutoDiff] Fix `@differentiable(linear)` type-checking.

dan-zheng opened the pull request:
For all non-`@noDerivative` parameter and result types, `@differentiable(linear)` function types should require and imply `T: Differentiable`, `T == T.TangentVector` requirements instead of `T: Differentiable & AdditiveArithmetic`.

Update tests.

---

Type-checking example:

```swift
// linear.swift
struct D: Differentiable {}
// `D` does not conform to `AdditiveArithmetic`.
// `D` is not equal to `D.TangentVector`.
func test(_: @differentiable(linear) (D) -> D) {}
```

Before:
```console
$ swift linear.swift
# Missing error.
```

After:
```console
$ swift linear.swift
linear.swift:5:39: error: parameter type 'D' does not conform to 'Differentiable' and satisfy 'D == D.TangentVector', but the enclosing function type is '@differentiable(linear)'; did you want to add '@noDerivative' to this parameter?
func test(_: @differentiable(linear) (D) -> D) {}
                                      ^
                                      @noDerivative
linear.swift:5:45: error: result type 'D' does not conform to 'Differentiable' and satisfy 'D == D.TangentVector', but the enclosing function type is '@differentiable(linear)'
func test(_: @differentiable(linear) (D) -> D) {}
                                            ^
```

---

Inference example:

```swift
func inferred<T, U>(_: @differentiable(linear) (T) -> U) {}
```

Before:
```console
$ swiftc -print-ast linear.swift
internal func inferred<T, U>(_: @differentiable(linear) (T) -> U) where T : AdditiveArithmetic, T : Differentiable, U : AdditiveArithmetic, U : Differentiable
```

After:
```console
$ swiftc -print-ast linear.swift
internal func inferred<T, U>(_: @differentiable(linear) (T) -> U) where T : Differentiable, T == T.TangentVector, U : Differentiable, U == U.TangentVector
```

dan-zheng committed:
[AutoDiff] Fix `@differentiable(linear)` type-checking.

For all non-`@noDerivative` parameter and result types, `@differentiable(linear)`
function types should require and imply `T: Differentiable`, `T == T.TangentVector`
requirements instead of `T: Differentiable & AdditiveArithmetic`.

Update tests.

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented:
> ```console
> linear.swift:5:39: error: parameter type 'D' does not conform to 'Differentiable' and satisfy 'D == D.TangentVector', but the enclosing function type is '@differentiable(linear)'; did you want to add '@noDerivative' to this parameter?
> func test(_: @differentiable(linear) (D) -> D) {}
>                                       ^
>                                       @noDerivative
> linear.swift:5:45: error: result type 'D' does not conform to 'Differentiable' and satisfy 'D == D.TangentVector', but the enclosing function type is '@differentiable(linear)'
> func test(_: @differentiable(linear) (D) -> D) {}
>                                             ^
> ```

Two issues:
* Why are there duplicate errors?
* The fixit is not appropriate the there's only one argument, because the function will still be illformed when you add `@noDerivative`.

rxwei commented on lib/AST/GenericSignatureBuilder.cpp:
```diff
@@ -5073,31 +5073,47 @@ class GenericSignatureBuilder::InferRequirementsWalker : public TypeWalker {
+        auto addRequirements = [&](Type type, bool isLinear) {
+          addConformanceConstraint(type, differentiableProtocol);
+          if (isLinear)
+            addSameTypeConstraint(type, tangentVectorAssocType);
```
If you add `T.TangentVector == Self`, then you don't need to add `T: AdditiveArithmetic` anymore I believe.

dan-zheng replied:
`T: AdditiveArithmetic` isn't added anymore.

dan-zheng commented:
> * Why are there duplicate errors?

The errors are not duplicate: one's is for the invalid parameter and one's for the invalid result.

> * The fixit is not appropriate the there's only one argument, because the function will still be illformed when you add `@noDerivative`.

This was the existing behavior. We could fix the fix-it so it appears only if at least one differentiability parameter remains after adding `@noDerivative`. I'd prefer to do this in a follow-up for incremental development.

rxwei reviewed (approved):
Not adding `T.TangentVector == T` was by design, but adding this constraint is okay with me. Did you encounter any concrete use cases where the absence of `T.TangentVector == T` requires extra user attention?

dan-zheng commented:
> Not adding `T.TangentVector == T` was by design, but adding this constraint is okay with me. Did you encounter any concrete use cases where the absence of `T.TangentVector == T` requires extra user attention?

Ah, interesting. I assumed that `T == T.TangentVector` is necessary since `@transpose` attribute type-checking requires it for linearity parameters - this patch makes `@differentiable(linear)` type-checking consistent.

Do we want to consistently require `T: Differentiable, T == T.TangentVector` for `@transpose` and `@differentiable(linear)` function types, or to relax it to `T: Differentiable & AdditiveArithmetic`?

dan-zheng commented:
Consensus after chatting: let's consistently require `T == T.TangentVector` for now.

dan-zheng committed:
Emit `@noDerivative` fixit only when valid.

Emit `@noDerivative` fixit only when there is at least one valid
differentiability/linearity parameter. Otherwise, adding `@noDerivative`
results in an ill-formed `@differentiable` function type.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
